testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/user/profile/DynamicUserProfileProvider.java

```diff
@@ -0,0 +1,213 @@
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.userprofile.UserProfileContext;
+import org.keycloak.userprofile.UserProfileMetadata;
+import org.keycloak.userprofile.AttributeMetadata;
+import org.keycloak.userprofile.UserProfileProvider;
+import org.keycloak.userprofile.legacy.AbstractUserProfileProvider;
+import org.keycloak.userprofile.AttributeValidatorMetadata;
+import org.keycloak.util.JsonSerialization;
+
+/**
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class DynamicUserProfileProvider extends AbstractUserProfileProvider<DynamicUserProfileProvider>
+        implements AmphibianProviderFactory<DynamicUserProfileProvider> {
+
+    public static final String ID = "dynamic-userprofile-provider";
+    private static Map<String, Map<UserProfileContext, UserProfileMetadata>> METADATA_CACHE;
+
+    public DynamicUserProfileProvider() {
+        // for reflection
+    }
+
+    public DynamicUserProfileProvider(KeycloakSession session, Map<UserProfileContext, UserProfileMetadata> metadataRegistry) {
+        super(session, metadataRegistry);
+    }
+
+    @Override
+    public String getId() {
+        return ID;
+    }
+
+    @Override
+    protected DynamicUserProfileProvider create(KeycloakSession session, Map<UserProfileContext, UserProfileMetadata> metadataRegistry) {
+        return new DynamicUserProfileProvider(session, metadataRegistry);
+    }
+
+    @Override
+    protected UserProfileMetadata configureUserProfile(UserProfileMetadata metadata) {
+        // adds an attribute config for any context
+        metadata.addAttribute("address", createStaticValueValidator("invalidAddress", "fixed-address"));
+
+        // adds an attribute config where the config only takes effect when a scope is requested
+        // this only works when performing actions during authentication
+        metadata.addAttribute("business.address", Arrays.asList("customer"),
+                createStaticValueValidator("invalidBusinessAddress", "fixed-business-address"));
+
+        if (UserProfileContext.ACCOUNT.equals(metadata.getContext())) {
+            // department not mandatory when in account and only readonly
+            metadata.addAttribute("department", createRequiredValidator("departmentRequired"), true, true);
+        }
+
+        if (UserProfileContext.USER_API.equals(metadata.getContext())) {
+            // department is set by admin and mandatory
+            metadata.addAttribute("department", createRequiredValidator("departmentRequired"));
+        }
+
+        return metadata;
+    }
+
+    @Override
+    protected UserProfileMetadata configureUserProfile(UserProfileMetadata metadata, KeycloakSession session) {
+        return getRealmUserProfile(session).computeIfAbsent(metadata.getContext(), context -> decorateUserProfile(metadata, session));
```

Comment:
If I understand this correctly, there are separate UserProfileMetadata instances for each context in each realm - they are cached so they are loaded only once and then reused until realm is reconfigured. 
Inside of the UserProfileMetadata instance I can  setup attributes per scopes. 
But I'm not sure if this mechanism has enough expresivity to support our configurations per client - where/how are we going to select user profile based on calling client with evaluation of the defaults chain?

I'm also probably missing when is `protected UserProfileMetadata configureUserProfile(UserProfileMetadata metadata)` method called and what is its difference from this method (which should be used for what)

Reply:
Attributes have a selection mechanism based on a `Predicate`. So you should be able to have any check for that.

The two versions of a `configureUserProfile` are for `init` and `runtime`. As per doc as well as the additional `KeycloakSession` parameter that is only used when at runtime, processing requests.

Reply:
Ah, OK, `Predicate`, I forgot this possibility. Thanks